File: jetbrains/src/main/kotlin/identitybraces/editor/BraceMapCache.kt
package identitybraces.editor

import com.intellij.openapi.editor.Document
import com.intellij.openapi.util.Key
import com.intellij.openapi.util.UserDataHolderEx
import identitybraces.core.BraceMap
import identitybraces.core.BraceScanner
import identitybraces.settings.IdentityBracesSettings

/**
 * Caches the brace map for the most recent state of one document.
 *
 * Every editor on a document, the hover, the guides and the scene director all derive their
 * answers from this, so they stay in agreement without talking to each other: given the same
 * text and the same settings, [BraceScanner] is a pure function.
 *
 * Keyed on the document's modification stamp and the settings version. Typing produces a
 * new stamp per keystroke and therefore a full rescan; that is a linear pass over the file —
 * about 20 ms per megabyte — which is why `maxFileLength` exists.
 */
class BraceMapCache private constructor() {
    private val gate = Any()

    private var stamp: Long = -1
    private var settingsVersion: Int = -1
    private var map: BraceMap = BraceMap.EMPTY

    /** The map for the document's current text, scanning only if something changed. */
    fun get(document: Document): BraceMap {
        val settings = IdentityBracesSettings.instance
        val currentStamp = document.modificationStamp
        val currentVersion = settings.version

        synchronized(gate) {
            if (stamp == currentStamp && settingsVersion == currentVersion) {
                return map
            }
        }

        // Scan outside the lock. Two threads may briefly duplicate the work; that is cheaper
        // than serialising every reader behind a multi-millisecond scan.
        val fresh = if (!settings.enabled || document.textLength > settings.maxFileLength) {
            BraceMap.EMPTY
        } else {
            BraceMap(BraceScanner.scan(document.immutableCharSequence, settings.toScanSettings()))
        }

        synchronized(gate) {
            stamp = currentStamp
            settingsVersion = currentVersion
            map = fresh
        }

        return fresh
    }

    /** The cached map if it is current, without scanning. Null when stale. */
    fun peek(document: Document): BraceMap? {
        synchronized(gate) {
            return if (stamp == document.modificationStamp && settingsVersion == IdentityBracesSettings.instance.version) map else null
        }
    }

    companion object {
        private val KEY = Key.create<BraceMapCache>("identitybraces.map")

        fun of(document: Document): BraceMapCache {
            return (document as UserDataHolderEx).putUserDataIfAbsent(KEY, BraceMapCache())
        }
    }
}

File: src/IdentityBraces/Core/BraceMapCache.cs
using IdentityBraces.Options;
using Microsoft.VisualStudio.Text;

namespace IdentityBraces.Core
{
    /// <summary>
    /// Caches the brace map for the most recent snapshot of one buffer.
    /// </summary>
    /// <remarks>
    /// Both the tagger and the adornment manager derive their answers from this, so they
    /// stay in agreement without talking to each other: given the same snapshot and the
    /// same settings, <see cref="BraceScanner"/> is a pure function.
    /// <para>
    /// A snapshot is immutable, so reference equality is a sound cache key. Typing produces
    /// a new snapshot per keystroke and therefore a full rescan; that is a linear pass over
    /// the file, which is why <see cref="IdentityBracesSettings.MaxFileLength"/> exists.
    /// </para>
    /// </remarks>
    internal sealed class BraceMapCache
    {
        private readonly object _gate = new object();

        // Two entries, not one. The tagger, the adornment manager and the line transform
        // source all read this, and during an edit they can briefly be looking at different
        // snapshots. With a single slot those callers evict each other on every access and
        // the file is rescanned per call instead of per change — which in the layout pass
        // means a full rescan for every visible line.
        private ITextSnapshot _snapshot;
        private BraceMap _map = BraceMap.Empty;
        private ITextSnapshot _previousSnapshot;
        private BraceMap _previousMap = BraceMap.Empty;
        private int _settingsVersion = -1;

        public static BraceMapCache GetOrCreate(ITextBuffer buffer)
        {
            return buffer.Properties.GetOrCreateSingletonProperty(
                typeof(BraceMapCache),
                delegate { return new BraceMapCache(); });
        }

        public BraceMap Get(ITextSnapshot snapshot)
        {
            IdentityBracesSettings settings = IdentityBracesSettings.Current;

            lock (_gate)
            {
                if (_settingsVersion == settings.Version)
                {
                    if (ReferenceEquals(_snapshot, snapshot))
                    {
                        return _map;
                    }

                    if (ReferenceEquals(_previousSnapshot, snapshot))
                    {
                        return _previousMap;
                    }
                }
            }

            // Scan outside the lock. Two threads may briefly duplicate the work; that is
            // cheaper than serialising every reader behind a multi-millisecond scan.
            BraceMap map;
            if (!settings.Enabled || snapshot.Length > settings.MaxFileLength)
            {
                map = BraceMap.Empty;
            }
            else
            {
                map = new BraceMap(BraceScanner.Scan(snapshot.GetText(), settings.ToScanSettings()));
            }

            lock (_gate)
            {
                if (_settingsVersion != settings.Version)
                {
                    // Settings changed: every cached map is stale, not just the older one.
                    _previousSnapshot = null;
                    _previousMap = BraceMap.Empty;
                }
                else
                {
                    _previousSnapshot = _snapshot;
                    _previousMap = _map;
                }

                _snapshot = snapshot;
                _map = map;
                _settingsVersion = settings.Version;
            }

            return map;
        }
    }
}
